### ISSUE TITLE
G2 and G3 Gcodes not displayed in gui

### ISSUE THREAD

**Issue by aduy** [UnfinishedBusiness/ncPilot#3]
It looks like it just skips drawing it because it doesnt know how to interpret the g2 or g3 gcodes.

heres my code:

G90 G94
G17
G20

(Operation 1 of 1 : 2D Profile2  2 )

G90
G54
G0 X3.298 Y0.2245
Z0.6
X3.298 Y0.2245
fire_torch
G1 X3.1581 Y0.086 F30
G17 G2 X3.0686 Y-0.03 Z0 I-0.0895 J-0.0235
G1 X0.0625 Y-0.03
G2 X-0.03 Y0.0625 Z0 I0 J0.0925
G1 X-0.03 Y1.4375
G2 X0.0625 Y1.53 Z0 I0.0925 J0
G1 X0.22 Y1.53
X0.22 Y2.97
X0.0625 Y2.97
G2 X-0.03 Y3.0625 Z0 I0 J0.0925
G1 X-0.03 Y4.4375
X-0.03 Y4.4375
G2 X0.0625 Y4.53 Z0 I0.0925 J0
G1 X0.22 Y4.53
X0.22 Y5.9375
X0.22 Y5.9375
G2 X0.3125 Y6.03 Z0 I0.0925 J0
G1 X1.5266 Y6.03
G2 X1.6161 Y5.961 Z0 I0 J-0.0925
G1 X1.6161 Y5.961
X3.1581 Y0.086
X3.3479 Y0.0341
torch_off
G0 X3.3479 Y0.0341 Z0.6

M5
G0 X0 Y0
M30
%

and heres a screenshot from another gcode sender I use.
![image](https://user-images.githubusercontent.com/2220585/81916798-2a3e3700-9589-11ea-98b2-68322d9fe9c8.png)

and here is a screenshot in the ncpilot.
![image](https://user-images.githubusercontent.com/2220585/81916676-f5ca7b00-9588-11ea-9a5b-0dee0a037836.png)

**Comment by aduy** [comment]
I looked at your sheetcam post. looks like you only use G0 G1, fire_torch and torch_off. Also best I can tell ncpilot requires a G1 or G0 for every line, its non modal, like most 3d printers.

I'm used to my big old haas cnc machine that uses a usb floppy emulator, and has 1mb of onboard storage so the code has to be efficient, or use multiple "floppies". In this case though its a non issue, the many segmented point to point arcs should be fine.

I primarily use fusion 360, so i have already altered a grbl post to only use your supported gcodes, still need to make it output fire_torch for m3 and torch_off for m5.
```

G0 X3.298 Y0.2245
G0 Z0.6
fire_torch
G0 X3.298 Y0.2245
G0 X3.298 Y0.2245 Z0
G1 X3.1581 Y0.086 F30
G1 X3.1593 Y0.0808
G1 X3.1601 Y0.0756
G1 X3.1608 Y0.0704
G1 X3.1611 Y0.0651
G1 X3.161 Y0.0598
G1 X3.1607 Y0.0546
G1 X3.1601 Y0.0493
G1 X3.1592 Y0.0441
G1 X3.158 Y0.039
G1 X3.1566 Y0.0339
G1 X3.1548 Y0.0289
G1 X3.1527 Y0.0241
G1 X3.1504 Y0.0193
G1 X3.1478 Y0.0147
G1 X3.1449 Y0.0103
G1 X3.1418 Y0.006
G1 X3.1385 Y0.0019
G1 X3.1349 Y-0.002
G1 X3.1311 Y-0.0057
G1 X3.1271 Y-0.0091
G1 X3.1229 Y-0.0123
G1 X3.1186 Y-0.0153
G1 X3.1141 Y-0.018
G1 X3.1094 Y-0.0205
G1 X3.1046 Y-0.0227
G1 X3.0997 Y-0.0246
G1 X3.0946 Y-0.0263
G1 X3.0895 Y-0.0276
G1 X3.0844 Y-0.0286
G1 X3.0791 Y-0.0294
G1 X3.0739 Y-0.0298
G1 X3.0686 Y-0.03
G1 X0.0625 Y-0.03
G1 X0.0571 Y-0.0298
G1 X0.0518 Y-0.0294
G1 X0.0464 Y-0.0286
G1 X0.0412 Y-0.0275
G1 X0.036 Y-0.0261
G1 X0.0309 Y-0.0244
G1 X0.0259 Y-0.0224
G1 X0.021 Y-0.0202
G1 X0.0162 Y-0.0176
G1 X0.0117 Y-0.0148
G1 X0.0073 Y-0.0117
G1 X0.003 Y-0.0084
G1 X-0.001 Y-0.0048
G1 X-0.0048 Y-0.001
G1 X-0.0084 Y0.003
G1 X-0.0117 Y0.0073
G1 X-0.0148 Y0.0117
G1 X-0.0176 Y0.0162
G1 X-0.0202 Y0.021
G1 X-0.0224 Y0.0259
G1 X-0.0244 Y0.0309
G1 X-0.0261 Y0.036
G1 X-0.0275 Y0.0412
G1 X-0.0286 Y0.0464
G1 X-0.0294 Y0.0518
G1 X-0.0298 Y0.0571
G1 X-0.03 Y0.0625
G1 X-0.03 Y1.4375
G1 X-0.0298 Y1.4429
G1 X-0.0294 Y1.4482
G1 X-0.0286 Y1.4536
G1 X-0.0275 Y1.4588
G1 X-0.0261 Y1.464
G1 X-0.0244 Y1.4691
G1 X-0.0224 Y1.4741
G1 X-0.0202 Y1.479
G1 X-0.0176 Y1.4838
G1 X-0.0148 Y1.4883
G1 X-0.0117 Y1.4927
G1 X-0.0084 Y1.497
G1 X-0.0048 Y1.501
G1 X-0.001 Y1.5048
G1 X0.003 Y1.5084
G1 X0.0073 Y1.5117
G1 X0.0117 Y1.5148
G1 X0.0163 Y1.5176
G1 X0.021 Y1.5202
G1 X0.0259 Y1.5224
G1 X0.0309 Y1.5244
G1 X0.036 Y1.5261
G1 X0.0412 Y1.5275
G1 X0.0464 Y1.5286
G1 X0.0518 Y1.5294
G1 X0.0571 Y1.5298
G1 X0.0625 Y1.53
G1 X0.22 Y1.53
G1 X0.22 Y2.97
G1 X0.0625 Y2.97
G1 X0.0571 Y2.9702
G1 X0.0518 Y2.9706
G1 X0.0464 Y2.9714
G1 X0.0412 Y2.9725
G1 X0.036 Y2.9739
G1 X0.0309 Y2.9756
G1 X0.0259 Y2.9776
G1 X0.021 Y2.9798
G1 X0.0163 Y2.9824
G1 X0.0117 Y2.9852
G1 X0.0073 Y2.9883
G1 X0.003 Y2.9916
G1 X-0.001 Y2.9952
G1 X-0.0048 Y2.999
G1 X-0.0084 Y3.003
G1 X-0.0117 Y3.0073
G1 X-0.0148 Y3.0117
G1 X-0.0176 Y3.0162
G1 X-0.0202 Y3.021
G1 X-0.0224 Y3.0259
G1 X-0.0244 Y3.0309
G1 X-0.0261 Y3.036
G1 X-0.0275 Y3.0412
G1 X-0.0286 Y3.0464
G1 X-0.0294 Y3.0518
G1 X-0.0298 Y3.0571
G1 X-0.03 Y3.0625
G1 X-0.03 Y4.4375
G1 X-0.03 Y4.4375
G1 X-0.0298 Y4.4429
G1 X-0.0294 Y4.4483
G1 X-0.0286 Y4.4536
G1 X-0.0275 Y4.4589
G1 X-0.0261 Y4.464
G1 X-0.0244 Y4.4692
G1 X-0.0224 Y4.4742
G1 X-0.0202 Y4.479
G1 X-0.0176 Y4.4838
G1 X-0.0148 Y4.4883
G1 X-0.0117 Y4.4928
G1 X-0.0084 Y4.497
G1 X-0.0048 Y4.501
G1 X-0.001 Y4.5048
G1 X0.003 Y4.5084
G1 X0.0073 Y4.5117
G1 X0.0117 Y4.5148
G1 X0.0163 Y4.5176
G1 X0.021 Y4.5202
G1 X0.0259 Y4.5224
G1 X0.0309 Y4.5244
G1 X0.036 Y4.5261
G1 X0.0412 Y4.5275
G1 X0.0464 Y4.5286
G1 X0.0518 Y4.5294
G1 X0.0571 Y4.5298
G1 X0.0625 Y4.53
G1 X0.22 Y4.53
G1 X0.22 Y5.9375
G1 X0.22 Y5.9375
G1 X0.2202 Y5.9429
G1 X0.2206 Y5.9483
G1 X0.2214 Y5.9536
G1 X0.2225 Y5.9589
G1 X0.2239 Y5.964
G1 X0.2256 Y5.9692
G1 X0.2276 Y5.9742
G1 X0.2298 Y5.979
G1 X0.2324 Y5.9838
G1 X0.2352 Y5.9883
G1 X0.2383 Y5.9928
G1 X0.2416 Y5.997
G1 X0.2452 Y6.001
G1 X0.249 Y6.0048
G1 X0.253 Y6.0084
G1 X0.2573 Y6.0117
G1 X0.2617 Y6.0148
G1 X0.2663 Y6.0176
G1 X0.271 Y6.0202
G1 X0.2759 Y6.0224
G1 X0.2809 Y6.0244
G1 X0.286 Y6.0261
G1 X0.2912 Y6.0275
G1 X0.2964 Y6.0286
G1 X0.3018 Y6.0294
G1 X0.3071 Y6.0298
G1 X0.3125 Y6.03
G1 X1.5266 Y6.03
G1 X1.5319 Y6.0299
G1 X1.5371 Y6.0294
G1 X1.5424 Y6.0286
G1 X1.5475 Y6.0276
G1 X1.5527 Y6.0263
G1 X1.5577 Y6.0246
G1 X1.5626 Y6.0227
G1 X1.5674 Y6.0205
G1 X1.5721 Y6.018
G1 X1.5766 Y6.0153
G1 X1.581 Y6.0123
G1 X1.5852 Y6.0091
G1 X1.5891 Y6.0056
G1 X1.5929 Y6.002
G1 X1.5965 Y5.9981
G1 X1.5999 Y5.994
G1 X1.603 Y5.9897
G1 X1.6058 Y5.9853
G1 X1.6084 Y5.9807
G1 X1.6107 Y5.9759
G1 X1.6128 Y5.9711
G1 X1.6146 Y5.9661
G1 X1.6161 Y5.961
G1 X1.6161 Y5.961
G1 X3.1581 Y0.086
G1 X3.3479 Y0.0341
G0 X3.3479 Y0.0341 Z0.6
torch_off

M5
G0 X0 Y0
M30
%

```

**Comment by UnfinishedBusiness** [comment]
Yeah, only rapids and line moves are rendered in the viewer. The controller does support arcs but I found instability where occasionally GRBL would whip out a whole circle someplace it’s not supposed to and this controller setup has no problem opening and running a Gcode file over 100mb...

The Gcode parser does support G0 and G1 as modal if I recall (its been some time since I wrote it) but I personally like the line by line layout. Also note that don’t use G20 or G21 unless you setup the machine parameters in MM units. I usually configure the machine parameters in inch units and never have any need to switch. I would also remove the z word from everything and let the fire_torch routine handle your Z height by probing. I always make sure that there’s noZ work offset set so that Z DRO and Z ABS match all the time.  The Z=0 button if for some of my customers using this control on a router
